enrichments/trace/internal/elastic/span.go

```diff
@@ -0,0 +1,294 @@
+	if isElasticTransaction(span) {
+		s.setTxnType(span)
+		s.setTxnResult(span)
+		s.setEventOutcome(span)
+	} else {
+		s.setEventOutcome(span)
+		s.setServiceTarget(span)
+	}
+}
+
+// normalizeAttributes sets any dependent attributes that
+// might not have been explicitly set as an attribute.
+func (s *Span) normalizeAttributes() {
+	if s.rpcSystem == "" && s.grpcStatus != "" {
+		s.rpcSystem = "grpc"
+	}
+}
+
+func (s *Span) setTxnType(span ptrace.Span) {
+	txnType := "unknown"
+	switch {
+	case s.isMessaging:
+		txnType = "messaging"
+		span.Attributes().PutStr("transaction.type", "messaging")
+	case s.isRPC, s.isHTTP:
+		txnType = "request"
+	}
+	span.Attributes().PutStr(AttributeTransactionType, txnType)
+}
+
+func (s *Span) setTxnResult(span ptrace.Span) {
+	var result string
+
+	if s.isHTTP && s.httpStatusCode > 0 {
+		switch i := s.httpStatusCode / 100; i {
+		case 1, 2, 3, 4, 5:
+			result = standardStatusCodeResults[i-1]
+		default:
+			result = fmt.Sprintf("HTTP %d", s.httpStatusCode)
+		}
+	}
+	if s.isRPC {
+		result = s.grpcStatus
+	}
+	if result == "" {
+		switch s.spanStatusCode {
+		case ptrace.StatusCodeError:
+			result = "Error"
+		default:
+			// default to success if all else fails
+			result = "Success"
+		}
+	}
+
+	span.Attributes().PutStr(AttributeTransactionResult, result)
+}
+
+func (s *Span) setEventOutcome(span ptrace.Span) {
+	// default to success outcome
+	outcome := "success"
+	switch {
+	case s.spanStatusCode == ptrace.StatusCodeError:
+		outcome = "failure"
+	case s.httpStatusCode >= http.StatusInternalServerError:
+		// TODO (lahsivjar): Handle GRPC status code? - not handled in apm-data
+		// TODO (lahsivjar): Move to HTTPResponseStatusCode? Backward compatibility?
+		outcome = "failure"
+	}
+	span.Attributes().PutStr(AttributeEventOutcome, outcome)
+}
+
+func (s *Span) setServiceTarget(span ptrace.Span) {
+	var targetType, targetName string
+
+	if s.peerService != "" {
+		targetName = s.peerService
+	}
+
+	switch {
+	case s.isDB:
+		targetType = "db"
+		if s.dbType != "" {
+			targetType = s.dbType
+		}
+		if s.dbName != "" {
+			targetName = s.dbName
+		}
+	case s.isMessaging:
+		targetType = "messaging"
+		if s.messagingSystem != "" {
+			targetType = s.messagingSystem
+		}
+		if !s.messagingDestnTemp && s.messagingDestn != "" {
+			targetName = s.messagingDestn
+		}
+	case s.isRPC:
+		targetType = "external"
+		if s.rpcSystem != "" {
+			targetType = s.rpcSystem
+		}
+		if s.rpcService != "" {
+			targetName = s.rpcService
+		}
+	case s.isHTTP:
+		targetType = "http"
+		if resource := getHostPort(s.urlFull, s.urlDomain, s.urlPort); resource != "" {
+			targetName = resource
+		}
+	}
+
+	if targetType != "" {
+		span.Attributes().PutStr(AttributeServiceTargetType, targetType)
+	}
+	if targetName != "" {
+		span.Attributes().PutStr(AttributeServiceTargetName, targetName)
+	}
+}
+
+func isElasticTransaction(span ptrace.Span) bool {
+	if span.ParentSpanID().IsEmpty() ||
```

Comment:
It would be different then what we do in apm-data right? Is it okay to create a follow-up for this?

Reply:
> We should ideally decide based on whether the span's parent is remote.

Hmm - very likely I miss something here, because I'm surprised by this. I thought this method would make the decision if this is a local transaction - so if this is the very first span (root span) in the given service. If there is no upstream service, then `isRemote` will still be `false`, but we'd still need to map that root span into a transaction.

> It would be different then what we do in apm-data right? I

Yeah, that's also [how I read it.](https://github.com/elastic/apm-data/blob/main/input/otlp/traces.go#L198)

Reply:
I have opened a followup PR for this: https://github.com/elastic/opentelemetry-lib/pull/48/commits/fe1da3a8821d5138cf65fdc6deec0ea24f3ff8d3

We can continue the discussion there.

Reply:
> If there is no upstream service, then isRemote will still be false, but we'd still need to map that root span into a transaction.

Sorry, clarifying: I meant instead of checking the span kind, rather than whether it's the root span.